src/coreclr/System.Private.CoreLib/src/System/Object.CoreCLR.cs

```diff
@@ -9,8 +10,14 @@ public partial class Object
     {
         // Returns a Type object which represent this object instance.
         [Intrinsic]
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        public extern Type GetType();
+        public unsafe Type GetType()
```

Comment:
First of all, it needs NoInlining on it. It's a jit intrinsic and if it gets inlined it might ruin some opts.
Also, it feels like it potentially doing more work prior calling runtime - are there any benchmarks? Presumably, GetType() is perf critical.

If you want to make it faster, you can inline the fast path here, something like:

```cs
nuint handle = RuntimeHelpers.GetMethodTable(o)->WriteableData->ExposedClassObject;
if (handle & 1)
    return (Type)(handle -1); // direct frozen RuntimeType object 
else
    return <fallback to internal call>
```

but last time I tried that - it was a perf regression so there only option I saw was to expand it in jit (https://github.com/dotnet/runtime/issues/88631)

Reply:
The helper is also doing unnecessary check for null and TypeDesc. With NoInling, the performance is likely to regress a lot.

It's not so clear about the desired approach - expanding in JIT or exposing more fields to managed code? Would GetHashCode also benefit from this?

Reply:
> The helper is also doing unnecessary check for null and TypeDesc. With NoInling, the performance is likely to regress a lot.

I don't think I follow, I meant NoInlining on top of `GetType()`

> Would GetHashCode also benefit from this?

`Object.GetHashCode` can also be optimized on C# side like this https://github.com/dotnet/runtime/pull/55273, it just needs a jit intrinsic to guarantee GC-safe access for object's header (exterior pointer). or can be expanded in JIT as well.

Reply:
> It's a jit intrinsic and if it gets inlined it might ruin some opts.

JIT intrinsics take precedence over inlining, no? Many JIT intrinsics are inlineable, but we do not have them marked as no inlining.

Reply:
> JIT intrinsics take precedence over inlining, no? Many JIT intrinsics are inlineable, but we do not have them marked as no inlining.

Nope, JIT is free to inline them. In 95% of cases it's not a problem because most intrinsic-related optimizations happen directly in importer prior inlining. There are only a few which are special and GetObject is one of those
E.g.

```
    case NI_System_Array_Clone:
    case NI_System_Collections_Generic_Comparer_get_Default:
    case NI_System_Collections_Generic_EqualityComparer_get_Default:
    case NI_System_Object_MemberwiseClone:
    case NI_System_Threading_Thread_get_CurrentThread:
    case NI_System_Object_GetType:
    case NI_System_Text_UTF8Encoding_UTF8EncodingSealed_ReadUtf8:
    case NI_System_SpanHelpers_SequenceEqual:
    case NI_System_Buffer_Memmove:
```
Some of them are just too big to be inlined anyway, internal calls or it's just not a big deal if jit inlines them.

Reply:
Can we disable the inlining for these in the JIT instead if the JIT does not want them inlined? 

Disabled inlining has secondary effect on code quality (forces stackwalkable frame, etc.). It will be hard to avoid perf regressions if the fast path has to be marked as disabled inlining.

Reply:
> Can we disable the inlining for these in the JIT instead if the JIT does not want them inlined?

Yeah, it can it be done